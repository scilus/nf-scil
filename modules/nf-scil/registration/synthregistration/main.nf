process REGISTRATION_SYNTHREGISTRATION {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_1.6.0.sif':
        'scilus/scilus:1.6.0' }"

    input:
    tuple val(meta), path(moving), path(fixed)


    output:
    tuple val(meta), path("*__*_output_warped.nii.gz"), emit: warped_image
    tuple val(meta), path("*__deform_warp.nii.gz"), emit: deform_transform
    tuple val (meta), path("*__init_warp.txt"), emit: affine_transform
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when_output_warped

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.suffix ?: ""

    def header = task.ext.header ? "-H "  + task.ext.header : ""
    def threads = task.ext.threads ? "-j "  + task.ext.threads : ""
    def gpu = task.ext.gpu ? "-g "  + task.ext.gpu : ""
    def smooth = task.ext.smooth ? "-s "  + task.ext.smooth : ""
    def extent = task.ext.extent ? "-e "  + task.ext.extent : ""
    def weight = task.ext.weight ? "-w "  + task.ext.weight : ""

    //For argument definition, mri_warp_convert -h
    def out = task.ext.out ? "--out" + task.ext.out : "--outlps"

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    mri_synthmorph -m affine -t ${prefix}__init_warp.txt $moving $fixed
    mri_synthmorph -m deform -i ${prefix}__init_warp.txt  -t temp.mgz -o ${prefix}__${suffix}_output_warped.nii.gz $moving $fixed

    mri_warp_convert -g $moving --inras temp.mgz $out ${prefix}__deform_warp.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        : \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' ))
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.suffix ?: "${meta.}"

    def header = task.ext.header ? "-H "  + task.ext.header : ""
    def threads = task.ext.threads ? "-j "  + task.ext.threads : ""
    def gpu = task.ext.gpu ? "-g "  + task.ext.gpu : ""
    def smooth = task.ext.smooth ? "-s "  + task.ext.smooth : ""
    def extent = task.ext.extent ? "-e "  + task.ext.extent : ""
    def weight = task.ext.weight ? "-w "  + task.ext.weight : ""

    def out = task.ext.out ? "--out" + task.ext.out : "lps"
    """
    mri_synthmorph -h

    mri_warp_convert -h

    touch ${prefix}__${suffix}_output_warped.nii.gz
    touch ${prefix}__deform_warp.nii.gz
    touch ${prefix}__init_warp.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        : \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' ))
    END_VERSIONS
    """
}
