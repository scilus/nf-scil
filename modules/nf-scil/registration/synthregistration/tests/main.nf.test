nextflow_process {

    name "Test Process REGISTRATION_SYNTHREGISTRATION"
    script "../main.nf"
    process "REGISTRATION_SYNTHREGISTRATION"

    tag "modules"
    tag "modules_nfcore"
    tag "registration"
    tag "registration/synthregistration"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "freesurfer.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("registration - synthregistration") {
        config "./nextflow.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                    [ id:'test', single_end:false ],
                    file("\${test_data_directory}/t1.nii.gz"),
                    file("\${test_data_directory}/fa.nii.gz")
                ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.warped_image.get(0).get(1)).name,
                    file(process.out.deform_transform.get(0).get(1)).name,
                    file(process.out.init_transform.get(0).get(1)).name
                ).match() }
            )
        }
    }
}
