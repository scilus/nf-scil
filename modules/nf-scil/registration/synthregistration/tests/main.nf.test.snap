{
    "registration - synthregistration": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output_warped.nii.gz:md5,28f54b497cda513feb86b07b0b28ac1d"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__deform_warp.nii.gz:md5,3cf3d49bc202410b8e6ef64ee394ad84"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__init_warp.lta:md5,46ad5a5d359b2ec157d1164135e746eb"
                    ]
                ],
                "3": [
                    "versions.yml:md5,49fb9c85da9f696926d1ab46ef0968fb"
                ],
                "deform_transform": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__deform_warp.nii.gz:md5,3cf3d49bc202410b8e6ef64ee394ad84"
                    ]
                ],
                "init_transform": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__init_warp.lta:md5,46ad5a5d359b2ec157d1164135e746eb"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,49fb9c85da9f696926d1ab46ef0968fb"
                ],
                "warped_image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output_warped.nii.gz:md5,28f54b497cda513feb86b07b0b28ac1d"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-06-05T18:41:40.417146"
    }
}